File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open System.Diagnostics.CodeAnalysis
open Hw4.Calculator

[<ExcludeFromCodeCoverage>]
type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg: string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ ->  InvalidOperationException() |> raise
    
let tryParse (s: string) =
    match System.Double.TryParse(s) with
    | true, value -> value
    | _ -> ArgumentException() |> raise

let parseCalcArguments(args : string[]) =
    match isArgLengthSupported args with
    | true -> {
              arg1 = tryParse args[0];
              arg2 = tryParse args[2];
              operation = parseOperation args[1]
              }
    | false -> ArgumentException() |> raise
    

File: Homework8/Hw8/Parser/Parser.cs
using System.Globalization;
using Hw8.Calculator;
namespace Hw8.Parser;

public class Parser : IParser
{
    private Operation ParseOperation(string arg)
    {
        return arg.ToLower() switch
        {
            "plus" or "+" => Operation.Plus,
            "minus" or "-" => Operation.Minus,
            "multiply" or "*" => Operation.Multiply,
            "divide" or "/" => Operation.Divide,
            _ => Operation.Invalid
        };
    }

    public bool IsArgsCountSupported(int argsCount) => argsCount == 3;

    public void ParseCalcArguments(out double val1, out Operation operation, out double val2, params string[] args)
    {
        if (!Double.TryParse(args[0], NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out val1) 
            || !Double.TryParse(args[2], NumberStyles.AllowDecimalPoint,
                CultureInfo.InvariantCulture, out  val2))
            throw new ArgumentException(Messages.InvalidNumberMessage);
        
        operation = ParseOperation(args[1]);
        if (operation == Operation.Invalid)
            throw new InvalidOperationException(Messages.InvalidOperationMessage);

        if (operation == Operation.Divide && val2 < Double.Epsilon)
            throw new DivideByZeroException(Messages.DivisionByZeroMessage);
    }
}